plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -165,6 +188,52 @@ lensResolveProvider _ ideState plId cl uri rd@(ResolveOne _ uid) = do
 lensResolveProvider _ _ _ _ _ rd = do
    throwError $ PluginInvalidParams (T.pack $ "Unexpected argument for lens resolve handler: " <> show rd)
 
+
+-- | Provide explicit imports in inlay hints.
+-- Applying textEdits can make the import explicit.
+-- There is currently no need to resolve inlay hints,
+-- as no tooltips or commands are provided in the label.
+inlayHintProvider :: Recorder (WithPriority Log) -> PluginMethodHandler IdeState 'Method_TextDocumentInlayHint
+inlayHintProvider _ state _ InlayHintParams {_textDocument = TextDocumentIdentifier {_uri}, _range = visibleRange} = do
+    isIHSupported <- liftIO $ isInlayHintsSupported state
+    if isIHSupported
+    then do
+        nfp <- getNormalizedFilePathE _uri
+        (ImportActionsResult {forLens, forResolve}, pm) <- runActionE "ImportActions" state $ useWithStaleE ImportActions nfp
+        let inlayHints = [ generateInlayHints newRange ie pm
+                         | (range, int) <- forLens
+                         , Just newRange <- [toCurrentRange pm range]
+                         , isSubrangeOf newRange visibleRange
+                         , Just ie <- [forResolve IM.!? int]]
+        pure $ InL inlayHints
+    -- When the client does not support inlay hints, fallback to the code lens,
+    -- so this is Null
+    else do pure $ InR Null
+  where
+    -- The appropriate and intended position for the hint hints to begin
+    -- is the end of the range for the code lens.
+    --   import Data.Char (isSpace)
+    --   |--- range ----|-- IH ---|
+    --                  |^-paddingLeft
+    --                  ^-_position
+    generateInlayHints :: Range -> ImportEdit -> PositionMapping -> InlayHint
+    generateInlayHints (Range _ end) ie pm =
+      InlayHint { _position = end
+                , _label = InL $ mkLabel ie
+                , _kind = Just InlayHintKind_Type -- for type annotations
+                , _textEdits = fmap singleton $ toTEdit pm ie
+                , _tooltip = Nothing
```

Comment:
I think we should definitely consider using this! For example, the tooltip for the minimal imports inlay hint could say "This is an import list for exactly the names used by the current module", or something. That way it's a bit more discoverable what it means. It's a nice UX feature!

Reply:
Could this be as simple as "Make this import explicit"? This doesn't even require resolve, even though the LSP spec recommends that the tooltip be obtained via resolve, but it's simple enough.

Reply:
Yes, I think either of those would be fine. I think we can just make it a constant string; the hint itself shows the things that are going to be inserted. So no need for resolve. Resolve is only for things that are expensive to compute.